File: Compare_NRP/CPLEX-For-NRP/CP/cpp_model/MiniCP/minicpp/examples/qap.cpp
/*
 * mini-cp is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License  v3
 * as published by the Free Software Foundation.
 *
 * mini-cp is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY.
 * See the GNU Lesser General Public License  for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with mini-cp. If not, see http://www.gnu.org/licenses/lgpl-3.0.en.html
 *
 * Copyright (c)  2018. by Laurent Michel, Pierre Schaus, Pascal Van Hentenryck
 */

#include <iostream>
#include <iomanip>
#include <fstream>
#include "solver.hpp"
#include "trailable.hpp"
#include "intvar.hpp"
#include "constraint.hpp"
#include "search.hpp"

int main(int argc,char* argv[])
{
    using namespace std;
    using namespace Factory;

    ifstream data("data/qap.txt");
    int n;
    data >> n;
    Matrix<int,2> w({n,n});
    for(int i =0;i < n;i++)
       for(int j=0;j < n;j++)
          data >> w[i][j];
    Matrix<int,2> d({n,n});
    for(int i =0;i < n;i++)
       for(int j=0;j < n;j++)
          data >> d[i][j];

    /*
    for(int i =0;i < n;i++) {
       for(int j=0;j < n;j++)
          cout << setw(3) << w[i][j] << " ";
       cout << endl;
    }
    for(int i =0;i < n;i++) {
       for(int j=0;j < n;j++)
          cout << setw(3) << d[i][j] << " ";
       cout << endl;
    }
    */

    
    
    CPSolver::Ptr cp  = Factory::makeSolver();
    auto x = Factory::intVarArray(cp,n,n);
    cp->post(Factory::allDifferent(x));
    auto wDist = Factory::intVarArray(cp,n*n);
    for(int k=0,i=0;i < n;i++)
       for(int j=0;j < n;j++)
          wDist[k++] = w[i][j] * Factory::element(d,x[i],x[j]);
    Objective::Ptr obj = Factory::minimize(Factory::sum(wDist));
       
    DFSearch search(cp,firstFail(cp,x));

    search.onSolution([&obj]() {
                         cout << "objective = " << obj->value() << endl;
                      });

    auto stat = search.optimize(obj);
    cout << stat << endl;    
    cp.dealloc();
    return 0;
}

File: Compare_NRP/CPLEX-For-NRP/CP/cpp_model/MiniCP/master/examples/pybench/qap.py
#!/usr/local/bin/python3
from minicpp import *

with open("examples/data/qap.txt") as f:
    n = int(f.readline())
    f.readline() # skip blank line
    data = [[int(v) for v in line.split()] for line in f.readlines()]
    w = data[:n]
    d = data[n+1:]
        
cp = makeSolver()
x  = intVarArray(cp,n,n)

cp.post(allDifferent(x))
wDist = [w[i][j] * element(d,x[i],x[j]) for i in range(0,n) for j in range(0,n)]

print(type(wDist))
print(wDist)

obj = minimize(sum(wDist))

print("Starting search...")
search = DFSearch(cp,firstFail(cp,x))
search.onSolution(lambda : print("objective = %d " % obj.value()))
stat = search.optimize(obj)
print(stat)


